File: apps/ios/MastermindTerminal/ShellBridge.swift
import Foundation
import UIKit
import WebKit

/// One row of the page's drawing-tool registry, as it arrives over the bridge.
///
/// `label` is the web's English label and `group` its TV-taxonomy family key — both are
/// the page's to define. Native only presents them: it localizes `label` through `L10n`
/// and buckets by `group`, and it must never filter, extend or reorder the inventory,
/// because the tools that exist are whatever the renderer in `terminal/` registers.
struct ShellDrawTool: Identifiable, Equatable {
    let id: String
    let label: String
    let group: String
}

/// Native side of bridge v1 (web types: terminal/lib/platform/contract.ts; message
/// schema: contracts/native-shell.v1.schema.json).
///
/// Web → native: the page posts {type: ...} objects to webkit.messageHandlers.mm.
/// Native → web: commands call window.__mmShell methods with JSON-encoded arguments.
/// WebKit delivers script messages on the main thread, so @Published writes are safe here.
final class ShellBridge: NSObject, ObservableObject, WKScriptMessageHandler {
    static let messageName = "mm"

    @Published private(set) var isReady = false
    @Published private(set) var symbol = AppConfig.defaultSymbol
    @Published private(set) var timeframe = ""
    /// R2.4 — the user's starred timeframes, which the interval wheel prefers over the
    /// canonical list. Published because the wheel has to rebuild when it arrives.
    @Published private(set) var favTimeframes: [String] = []
    /// R2.1 — the chart's own drawing-tool registry, sent by the page. Native never
    /// hardcodes an engine tool list (`AGENTS.md`: the registry lives in `terminal/`).
    @Published private(set) var drawTools: [ShellDrawTool] = []
    private(set) var availableTimeframes: [String] = []

    weak var webView: WKWebView?

    /// Called before a reload so the loading cover comes back until the page re-announces ready.
    func reset() {
        isReady = false
    }

    func userContentController(_ userContentController: WKUserContentController, didReceive message: WKScriptMessage) {
        guard message.name == Self.messageName,
              let body = message.body as? [String: Any],
              let type = body["type"] as? String else { return }
        switch type {
        case "ready":
            isReady = true
            availableTimeframes = (body["availableTimeframes"] as? [String]) ?? []
            applyOptionalPayload(body)
        case "symbolChanged":
            if let sym = body["sym"] as? String, !sym.isEmpty { symbol = sym }
        case "stateChanged":
            if let tf = body["tf"] as? String, !tf.isEmpty { timeframe = tf }
            applyOptionalPayload(body)
        case "openExternal":
            if let raw = body["url"] as? String, let url = URL(string: raw),
               url.scheme == "https" || url.scheme == "http" {
                UIApplication.shared.open(url)
            }
        default:
            break
        }
    }

    /// R2.4/R2.1 — the two payload members `ready` and `stateChanged` both carry. A key
    /// that is *absent* leaves the stored value alone: an older page build must not blank
    /// a list a newer one already delivered, and only the page can say the list is empty.
    private func applyOptionalPayload(_ body: [String: Any]) {
        if let favs = body["favTimeframes"] as? [String] { favTimeframes = favs }
        if let raw = body["drawTools"] as? [[String: Any]] { drawTools = Self.parseDrawTools(raw) }
    }

    private static func parseDrawTools(_ rows: [[String: Any]]) -> [ShellDrawTool] {
        rows.compactMap { row in
            guard let id = row["id"] as? String, !id.isEmpty,
                  let label = row["label"] as? String, !label.isEmpty else { return nil }
            return ShellDrawTool(id: id, label: label, group: (row["group"] as? String) ?? "")
        }
    }

    // MARK: - native → web commands

    func setSymbol(_ symbol: String) { call("setSymbol", args: [symbol]) }
    func setTimeframe(_ timeframe: String) { call("setTimeframe", args: [timeframe]) }
    func setLang(_ lang: String) { call("setLang", args: [lang]) }

    /// Hands the page a Supabase session. One object argument, matching the web's
    /// `setSession({access_token, refresh_token})`; the page reloads ITSELF once the
    /// cookies are written, so native must not reload after calling this.
    /// Fire-and-forget: the promise it returns is unreadable from here, and token
    /// values must never reach a log.
    func setSession(accessToken: String, refreshToken: String) {
        call("setSession", args: [["access_token": accessToken, "refresh_token": refreshToken]])
    }

    /// Shows/hides the web chart's own drawing toolbar. Legacy compat only (R2.1): the
    /// pencil now opens the native Drawings sheet and the web dock stays hidden in shell
    /// mode, so nothing in the app calls this. Kept because it is still bridge v1 surface.
    func setDrawTools(_ visible: Bool) {
        call("setDrawTools", args: [visible])
    }

    /// R2.1 — activates one drawing tool by its registry id (the same code path the web
    /// dock's own buttons take). Ids come from `drawTools`; native never invents one.
    func setDrawTool(_ id: String) { call("setDrawTool", args: [id]) }

    /// R2.1 — drawing history, owned by the chart renderer in `terminal/`.
    func drawUndo() { call("drawUndo", args: []) }
    func drawRedo() { call("drawRedo", args: []) }

    /// R2.2 — asks the page to open one of its own modals ("indicators" | "compare").
    /// One implementation law: the picker is the web's, presented by the web.
    func openPanel(_ id: String) { call("openPanel", args: [id]) }

    private func call(_ method: String, args: [Any]) {
        // Arguments are JSON-encoded and applied — never string-interpolated into JS.
        guard JSONSerialization.isValidJSONObject(args),
              let data = try? JSONSerialization.data(withJSONObject: args),
              let json = String(data: data, encoding: .utf8) else { return }
        let js = "window.__mmShell && window.__mmShell.\(method).apply(window.__mmShell, \(json));"
        webView?.evaluateJavaScript(js, completionHandler: nil)
    }
}

File: terminal/lib/platform/shellBridge.ts
"use client";
import {
  SHELL_BRIDGE_VERSION,
  type MmShellApi,
  type ShellDrawTool,
  type ShellOutboundMessage,
  type ShellPanelId,
} from "@/lib/platform/contract";
import { TF_CANONICAL_ORDER } from "@/lib/startTf";
import { applyLang } from "@/lib/i18n";

/** Post a notification to the native shell. Safe everywhere: without a native host it is a no-op. */
export function postToShell(msg: ShellOutboundMessage) {
  try {
    (window as unknown as { webkit?: { messageHandlers?: { mm?: { postMessage(m: unknown): void } } } })
      .webkit?.messageHandlers?.mm?.postMessage(msg);
  } catch { /* not hosted by a shell */ }
}

export type ShellBridgeHooks = {
  getState: () => { sym: string; tf: string };
  /** Extras that ride the `ready` notification (and every stateChanged post from the shell). */
  getPayload: () => { favTimeframes: string[]; drawTools: readonly ShellDrawTool[] };
  /** R2.5 commands — each reuses the web's own code path, never a second implementation. */
  setDrawTool: (id: string) => boolean;
  drawUndo: () => boolean;
  drawRedo: () => boolean;
  openPanel: (id: ShellPanelId) => boolean;
};

const PANEL_IDS: readonly ShellPanelId[] = ["indicators", "compare"];

/**
 * Install `window.__mmShell` and announce `ready`. Commands deliberately ride the SAME seams
 * the product already exposes — `mm:embedded-symbol` (dashboard warm-iframe symbol switch) and
 * `mm:set-tf` (TF toolbar) — so the bridge adds no second symbol/TF pathway to maintain.
 * Returns the uninstaller (used on unmount / shell-mode teardown).
 */
export function initShellBridge(hooks: ShellBridgeHooks): () => void {
  const api: MmShellApi = {
    version: SHELL_BRIDGE_VERSION,
    setSymbol: (sym) => {
      const s = String(sym || "").trim().toUpperCase();
      if (s) window.dispatchEvent(new CustomEvent("mm:embedded-symbol", { detail: { symbol: s } }));
    },
    setTimeframe: (tf) => {
      const t = String(tf || "").trim();
      if (t) window.dispatchEvent(new CustomEvent("mm:set-tf", { detail: { tf: t } }));
    },
    setLang: (lang) => applyLang(lang === "zh" ? "zh" : "en"),
    setDrawTools: (visible) => {
      // Shell mode hides the drawing toolbar by default (chart real estate); the native
      // pencil toggles it back in — pure CSS class flip, no drawing state touched.
      document.querySelector(".app")?.classList.toggle("shell-draw", !!visible);
      return true;
    },
    setDrawTool: (id) => hooks.setDrawTool(String(id || "")),
    drawUndo: () => hooks.drawUndo(),
    drawRedo: () => hooks.drawRedo(),
    openPanel: (id) => (PANEL_IDS.includes(id as ShellPanelId) ? hooks.openPanel(id as ShellPanelId) : false),
    restoreState: ({ sym, tf } = {}) => {
      if (sym) api.setSymbol(sym);
      if (tf) api.setTimeframe(tf);
    },
    setSession: async ({ access_token, refresh_token }) => {
      if (!access_token || !refresh_token) return false;
      const { createClient } = await import("@/lib/supabase/client");
      const { error } = await createClient().auth.setSession({ access_token, refresh_token });
      // Reload so server components and route handlers see the freshly written auth cookies.
      if (!error) setTimeout(() => window.location.reload(), 80);
      return !error;
    },
    getState: () => hooks.getState(),
  };
  (window as unknown as { __mmShell?: MmShellApi }).__mmShell = api;
  const payload = hooks.getPayload();
  postToShell({
    type: "ready",
    bridgeVersion: SHELL_BRIDGE_VERSION,
    availableTimeframes: [...TF_CANONICAL_ORDER],
    favTimeframes: [...payload.favTimeframes],
    drawTools: [...payload.drawTools],
  });
  return () => {
    const w = window as unknown as { __mmShell?: MmShellApi };
    if (w.__mmShell === api) w.__mmShell = undefined;
  };
}
